topdown/eval.go

```diff
@@ -2694,35 +2694,31 @@ func (e evalVirtualPartial) partialEvalSupportRule(rule *ast.Rule, path ast.Ref)
 		// Skip this rule body if it fails to type-check.
 		// Type-checking failure means the rule body will never succeed.
 		if e.e.compiler.PassesTypeCheck(plugged) {
-			var key, value *ast.Term
-
-			if rule.Head.Key != nil {
-				key = child.bindings.PlugNamespaced(rule.Head.Key, e.e.caller.bindings)
-			}
+			var value *ast.Term
 
 			if rule.Head.Value != nil {
 				value = child.bindings.PlugNamespaced(rule.Head.Value, e.e.caller.bindings)
 			}
 
-			head := ast.NewHead(rule.Head.Name, key, value)
+			ref := e.e.namespaceRef(rule.Ref())
+			for i := 1; i < len(ref); i++ {
+				ref[i] = child.bindings.plugNamespaced(ref[i], e.e.caller.bindings)
+			}
+			pkg, ruleRef := splitPackageAndRule(ref)
 
-			ruleRef := e.e.namespaceRef(rule.Ref())
-			l := len(path)
-			if l <= len(ruleRef) {
-				ruleRef = ruleRef[l-1:] // safe; path always has the data document prefix
-				if s, ok := ruleRef[0].Value.(ast.String); ok {
-					ruleRef[0].Value = ast.Var(s)
-				}
-				for i := 1; i < len(ruleRef); i++ {
-					ruleRef[i] = child.bindings.plugNamespaced(ruleRef[i], e.e.caller.bindings)
-				}
-				head.Reference = ruleRef
-				if head.Name.Equal(ast.Var("")) && (len(ruleRef) == 1 || (len(ruleRef) == 2 && head.Key == nil)) {
-					head.Name = ruleRef[0].Value.(ast.Var)
-				}
-				if len(ruleRef) == 2 && head.Key == nil {
-					head.Key = ruleRef[len(ruleRef)-1]
-				}
+			head := ast.RefHead(ruleRef, value)
+
+			// key is also part of ref in single-value rules, and can be dropped
+			if rule.Head.Key != nil && rule.Head.RuleKind() == ast.MultiValue {
+				head.Key = child.bindings.PlugNamespaced(rule.Head.Key, e.e.caller.bindings)
+			}
+
+			if rule.Head.RuleKind() == ast.SingleValue && len(ruleRef) == 2 {
+				head.Key = ruleRef[len(ruleRef)-1]
+			}
+
+			if head.Name.Equal(ast.Var("")) && (len(ruleRef) == 1 || (len(ruleRef) == 2 && rule.Head.RuleKind() == ast.SingleValue)) {
+				head.Name = ruleRef[0].Value.(ast.Var)
```

Comment:
So we're basically back-filling `Name` now for ref-rules?
```rego
p.q.r[x][y][z] := 1
```
would have name `p`?

Reply:
Only if the length is max 2. I.e. an old style "non-ref" partial object or set